### ISSUE TITLE
Refactor System.Linq

### ISSUE THREAD

**Issue by rafidka** [dotnet/corefx#1149]
Following @stephentoub comment here:

https://github.com/dotnet/corefx/issues/1143#issuecomment-81683842

I am filing this issue to track doing some refactoring in System.Linq. Currently, everything is in one file which has about 3.2k lines and around 20 classes and interfaces. It is good if this can be refactored to make the code easier to follow.

**Comment by ellismg** [comment]
Did you want to do this as part of your work on increasing code coverage?

**Comment by rafidka** [comment]
Yes, sure, I would love to. It is just that I am not a contributor yet (I hope you make me one), so I can't assign tasks to myself.

**Comment by ellismg** [comment]
@rafidka Thanks!  I marked the is "grabbed by community" since you're going to take care of it.  Please let me know if you change your mind.

**Comment by rafidka** [comment]
@stephentoub, just wanted to give you my opinion on this task. Mainly, I want to move classes to their own files rather than everything in one big file which makes it hard to follow. I don't see this breaking anything, and additionally, if we add the unit tests first, then we can be confident about this refactoring.
